File: src/main/java/org/willxu/algorithm/service/integer/MaximumDepthOfBinaryTree.java
/*
 * Copyright (C) 2021  Weiyang(Will) Xu
 * 
 * This file is part of algorithm.
 * Algorithm is free software: you can redistribute it and/or modify
 * it under the term of the GNU General Public License version 3 or
 * any later version, as specified in the readme.md file.
 */

package org.willxu.algorithm.service.integer;

import org.willxu.algorithm.provide.TreeNode;

public interface MaximumDepthOfBinaryTree {
    /**
     * Given the root of a binary tree, return its maximum depth.
     *
     * A binary tree's maximum depth is the number of nodes along the
     * number of nodes along the longest path from the root node down
     * to the farthest leaf node.
     *
     * @param root 0 <= root.size <= 10^4
     *             -100 <= Node.val <= 100
     * @return max depth
     */
    int maxDepth(TreeNode root);
}

File: src/leetcode/p0100_0199/p0104_maximum_depth_of_binary_tree/maximum_depth_of_binary_tree.py
from abc import ABC, abstractmethod
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class MaximumDepthOfBinaryTree(ABC):
    """
    Given the root of a binary tree, return its maximum depth.
    A binary tree's maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.
    """
    @abstractmethod
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        """
        Args:
            root:
                - The number of nodes in the tree is in the range [0, 10^4].
                - -100 <= Node.val <= 100
        """
        pass

class MaximumDepthOfBinaryTreeRecursive(MaximumDepthOfBinaryTree):
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        l = self.maxDepth(root.left)
        r = self.maxDepth(root.right)
        return max(l, r) + 1
